Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift

```diff
@@ -0,0 +1,528 @@
+
+  @inlinable
+  public func receiveEnd() {
+    self.interceptors.receive(.end)
+  }
+
+  @inlinable
+  public func receiveError(_ error: Error) {
+    self.handleError(error)
+    self.finish()
+  }
+
+  @inlinable
+  public func finish() {
+    switch self.state {
+    case .idle:
+      self.interceptors = nil
+      self.state = .completed
+
+    case let .observing(_, _, statusPromise):
+      statusPromise.fail(GRPCStatus(code: .unavailable, message: nil))
+
+    case .completed:
+      self.interceptors = nil
+    }
+  }
+
+  // MARK: - Interceptors to User Function
+
+  @inlinable
+  internal func receiveInterceptedPart(_ part: GRPCServerRequestPart<Request>) {
+    switch part {
+    case let .metadata(headers):
+      self.receiveInterceptedMetadata(headers)
+    case let .message(message):
+      self.receiveInterceptedMessage(message)
+    case .end:
+      self.receiveInterceptedEnd()
+    }
+  }
+
+  @inlinable
+  internal func receiveInterceptedMetadata(_ headers: HPACKHeaders) {
+    switch self.state {
+    case .idle:
+      let statusPromise: EventLoopPromise<GRPCStatus> = context.eventLoop.makePromise()
+
+      // Make a context to invoke the user handler with.
+      let context = GRPCAsyncServerCallContext(
+        headers: headers,
+        logger: self.context.logger,
+        userInfoRef: self.userInfoRef
+      )
+
+      // Create a request stream to pass to the user function and capture the
+      // handler in the updated state to allow us to produce more results.
+      let requestStream = GRPCAsyncStream<Request>(AsyncThrowingStream { continuation in
+        self.state = .observing(
+          { streamEvent in
+            switch streamEvent {
+            case let .message(request): continuation.yield(request)
+            case .end: continuation.finish()
+            }
+          },
+          context,
+          statusPromise
+        )
+      })
+
+      // Create a writer that the user function can use to pass back responses.
+      let responseStreamWriter = GRPCAsyncResponseStreamWriter(
+        context: context,
+        compressionIsEnabled: self.context.encoding.isEnabled,
+        sendResponse: self.interceptResponse(_:metadata:)
+      )
+
+      // Send response headers back via the interceptors.
+      self.interceptors.send(.metadata([:]), promise: nil)
+
+      // Register callbacks on the status future.
+      statusPromise.futureResult.whenComplete(self.userFunctionStatusResolved(_:))
+
+      // Spin up a task to call the async user handler.
+      self.task = statusPromise.completeWithTask {
+        // Check for cancellation before calling the user function.
+        // This could be the case if the RPC has been cancelled or had an error before this task
+        // has been scheduled.
+        guard !Task.isCancelled else {
+          throw CancellationError()
+        }
+
+        // Call the user function.
+        try await self.observer(requestStream, responseStreamWriter, context)
```

Comment:
We should make sure that we correctly handle (not necessarily here) the user function throwing `GRPCStatus.ok`

Reply:
Um, this is an interesting one.

I'm looking at the semantics of the existing handlers where the user function is expected to complete a promise. If they fail the promise with `.ok` then that is equivalent to the user function throwing `GRPCStatus.ok` in the async-await handler. 

It's my claim that the current implementation in this PR is semantically equivalent to the existing one. That is to say that it will end up calling `handleError(error, thrownFromHandler: true)` in both cases and it ends up in the common layer of `ServerErrorProcessor.processObserverError`.

So first off, do you agree that the PR currently implements the same semantics as the existing handler? If not, then I'll need to go away and look again.

And, if you do, are you saying you would like for us to have different semantics in the async-await–based API?

Reply:
I can at least add a test for this.

Reply:
Added `testHandlerThrowsGRPCStatusOK` in 2611a99d.

Reply:
> So first off, do you agree that the PR currently implements the same semantics as the existing handler? If not, then I'll need to go away and look again.

Sort of. This handler now does the job of four handlers and reviewing those revealed they have somewhat inconsistent semantics:
- unary: return a failed future of `GRPCStatus.ok` would result in no response message and an 'ok' status (seems incorrect to me)
- client streaming: 
  - caller fails stream-event handler callback with `.ok`; no response message and an 'ok' status (seems wrong)
  - caller fails response promise with `.ok`; no response message and an 'ok' status (seems wrong)
- server streaming: API requires user to return a `GRPCStatus` future; 'ok' is always fine.
- bidirectional streaming:
  - caller fails stream-event handler callback with '.ok': no messages but '.ok' sent back (seems wrong but perhaps much less so than in the client streaming case because empty response streams are allowed) 
  - user must complete a `GRPCStatus` promise; 'ok' is fine.
  
To sum up: when we expect a single response (unary and client streaming) a thrown 'ok' status is _weird_ because the client will throw when awaiting response but the status will be okay. For bidirectional and server streaming throwing an 'ok' is fine (albeit slightly weird) because we allow empty streams. Ideally `.ok` wouldn't be an error but that ship has sadly already sailed.

> And, if you do, are you saying you would like for us to have different semantics in the async-await–based API?

Yes, I think we should catch thrown `.ok` statuses and handle them appropriately. For unary and client streaming we should turn them into a different status, for server streaming and bidirectional streaming we should treat them as an early return from the user function.

Reply:
> Yes, I think we should catch thrown .ok statuses and handle them appropriately. For unary and client streaming we should turn them into a different status, ...

Should we use the protocol violation error (in this case, it would be the server that violated the protocol)?

> ... for server streaming and bidirectional streaming we should treat them as an early return from the user function.

So, turn this into a status `.ok` with no responses?

> Ideally .ok wouldn't be an error but that ship has sadly already sailed.

Could we explicitly catch `.ok` from the user handler and fail in all cases with a protocol violation, or somehow deprecate the use of throwing a `GRPCStatus` in favour of throwing something that doesn't include `.ok`, given that this is a new API?

Reply:
> > Yes, I think we should catch thrown .ok statuses and handle them appropriately. For unary and client streaming we should turn them into a different status, ...
> 
> Should we use the protocol violation error (in this case, it would be the server that violated the protocol)?

Not sure if this is protocol violation, it's really a user error. `.unknown` might be the right status code here (since it's really not clear what the caller meant if they threw `.ok`). 

> > ... for server streaming and bidirectional streaming we should treat them as an early return from the user function.
> 
> So, turn this into a status `.ok` with no responses?

Yes, I think that's probably fine. WDYT? (I keep changing my mind on this one.)

> > Ideally .ok wouldn't be an error but that ship has sadly already sailed.
> 
> Could we explicitly catch `.ok` from the user handler and fail in all cases with a protocol violation, or somehow deprecate the use of throwing a `GRPCStatus` in favour of throwing something that doesn't include `.ok`, given that this is a new API?

I'm not sure it's worth it. It's such a pervasive type that deprecating and replacing the whole type to handle this one case arguably isn't worth the toil. Moreover users can still use deprecated types.

There is something to be said for always catching `.ok` and treating them as invalid for all four types of RPC though. Perhaps that's better than handling it when we stream responses.

Reply:
> There is something to be said for always catching .ok and treating them as invalid for all four types of RPC though. Perhaps that's better than handling it when we stream responses.

I prefer the consistency of this if it's acceptable to you? We could always document it clearly.

Reply:
Yes that's fine by me!